File: python/src/problems101_200/find_peak_element.py
from abc import ABC, abstractmethod
from typing import override


class Problem(ABC):
    @abstractmethod
    def find_peak_element(self, nums: list[int]) -> int:
        pass


class Default(Problem):
    @override
    def find_peak_element(self, nums: list[int]) -> int:
        start: int = 0
        end: int = len(nums) - 1
        while start < end:
            if nums[start] > nums[end]:
                end -= 1
            else:
                start += 1
        return nums[end]


SOLUTIONS: list[Problem] = [Default()]

File: javascript/src/problems101_200/find-peak-element.js
class Problem {
  findPeakElement(nums) {
    throw new Error(`Not implemented (${nums}).`);
  }
}

class Default extends Problem {
  findPeakElement(nums) {
    let start = 0;
    let end = nums.length - 1;
    while (start < end) {
      if (nums[start] > nums[end]) {
        end--;
      } else {
        start++;
      }
    }
    return nums[end];
  }
}

const SOLUTIONS = [new Default()];
export default SOLUTIONS;
